### PR TITLE
[SPARK-48472][SQL] Enable reflect expressions with collated strings

### PR DESCRIPTION
**Pull request by mihailoale-db**
### What changes were proposed in this pull request?

Changes made in this pull request enable collation of strings in "reflect" expressions.

### Why are the changes needed?

Changes are bug fix which enable users to use feature mentioned above.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Using unit test which can be found in sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by mihailoale-db** [comment]
@mihailom-db @uros-db Can you review this please. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala [hunk 1]
```diff
@@ -59,11 +62,11 @@ import org.apache.spark.util.Utils
   since = "2.0.0",
   group = "misc_funcs")
 case class CallMethodViaReflection(
-      children: Seq[Expression],
-      failOnError: Boolean = true)
+                                    children: Seq[Expression],
+                                    failOnError: Boolean = true)
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala:66]: nit: style

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala [hunk 2]
```diff
@@ -177,7 +182,7 @@ case class CallMethodViaReflection(
   @transient private lazy val buffer = new Array[Object](argExprs.length)
 
   override protected def withNewChildrenInternal(
-    newChildren: IndexedSeq[Expression]): CallMethodViaReflection = copy(children = newChildren)
+                                                  newChildren: IndexedSeq[Expression]): CallMethodViaReflection = copy(children = newChildren)
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala:185]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala [hunk 3]
```diff
@@ -59,11 +62,11 @@ import org.apache.spark.util.Utils
...
 case class CallMethodViaReflection(
-      children: Seq[Expression],
-      failOnError: Boolean = true)
+                                    children: Seq[Expression],
+                                    failOnError: Boolean = true)
   extends Nondeterministic
-  with CodegenFallback
-  with QueryErrorsBase {
+    with CodegenFallback
+    with QueryErrorsBase {
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala:69]: nit: style as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala [hunk 4]
```diff
@@ -229,10 +234,13 @@ object CallMethodViaReflection {
       } else {
         // Argument type must match. That is, either the method's argument type matches one of the
         // acceptable types defined in typeMapping, or it is a super type of the acceptable types.
-        candidateTypes.zip(argTypes).forall { case (candidateType, argType) =>
+        candidateTypes.zip(argTypes).forall { case (candidateType, argType)
+          if !argType.isInstanceOf[StringType] =>
           typeMapping(argType).exists(candidateType.isAssignableFrom)
+        case _ => true
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala:240]:
```suggestion
        case _ => candidateType.isAssignableFrom(classOf[String])
```
I believe this should not be only true. This should be reproducible from passing something that is a string to a function that expects e.g. integer.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 5]
```diff
@@ -208,6 +208,50 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
     }
   }
 
+  test("[SPARK-48472] Enable reflect expressions with collated strings") {
```
Comment (mihailom-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:211]: lets move this to `CollationSQLExpressionsSuite` and follow the pattern there.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala (hunk 4 above)
Comment (mihailom-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala:240]: Also add test for this case

**Comment by uros-db** [comment]
@mihailoale-db I've seen some nit comments regarding code style, generally I recommend following these 2 guidelines:

- avoid unnecessary changes in the existing code (be careful if your IDE may sometimes automatically reformat code)
- for Scala code, follow the general rules outlined here: https://github.com/databricks/scala-style-guide

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 6]
```diff
@@ -208,6 +208,50 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
...
+    checkError(
+      exception = intercept[SparkException] {
+        sql("SELECT REFLECT('java.util.UUID', 'fromString'," +
+          " collate('a5cf6c42-0c85-418f-af6c-3e4e5b1328f2', 'utf8_binary'))=" +
+          " collate('a5cf6c42-0c85-418f-af6c-3e4e5b1328f2', 'utf8_BS');")
+      },
+      errorClass = "COLLATION_INVALID_NAME",
+      sqlState = "42704",
+      parameters = Map("collationName" -> "utf8_BS")
+    )
```
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:252]: let's only focus on the `reflect` expression (the scope of this PR)
we already have tests that check whether collation names are valid

so we can remove this, but we should maybe check what happens if arguments have two different _valid_ collations. The function may or may not enforce some collation type casting rules, but this should probably be reflected in our tests either way

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala [hunk 7]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (uros-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala:1]: avoid unnecessary changes

**Review by mihailomilosevic2001 — APPROVED** [review]
LGTM now, @cloud-fan could you review this?

**Review by uros-db — APPROVED** [review]
lgtm

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
